analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -30,10 +30,16 @@ internal sealed class Binary : BranchingProcessor<IBinaryOperationWrapper>
     protected override SymbolicConstraint BoolConstraintFromOperation(SymbolicContext context, IBinaryOperationWrapper operation) =>
         BinaryConstraint(operation.OperatorKind, context.State[operation.LeftOperand], context.State[operation.RightOperand]);
 
-    protected override ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper operation, bool falseBranch) =>
-        operation.OperatorKind.IsAnyEquality()
-            ? LearnBranchingConstraint<ObjectConstraint>(state, operation, falseBranch) ?? LearnBranchingConstraint<BoolConstraint>(state, operation, falseBranch) ?? state
-            : state;
+    protected override ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper operation, bool falseBranch)
+    {
+        if (operation.OperatorKind.IsAnyEquality())
+        {
+            state = LearnBranchingConstraint<ObjectConstraint>(state, operation, falseBranch) ?? state;
+            state = LearnBranchingConstraint<BoolConstraint>(state, operation, falseBranch) ?? state;
+            return state;
```

Comment:
We now have the combination of `BoolTrue`/`BoolFalse` and `NotNull` that need to be learned from so we can not short-circuit anymore.

Reply:
This should be in another PR to clearly see the impact on UTs